docs/semgrepignore-v2-reference.md

```diff
@@ -0,0 +1,164 @@
+# Semgrepignore v2 reference
+
+This document covers the Semgrepignore **v2** target filtering system that is
+currently available with the `--experimental` option of the `semgrep`
+command. It differs from the legacy **v1** implementation.
+referred to as "v1".
+
+## The target filtering process
+
+A `semgrep scan` command takes one or more scanning roots as
+arguments. The default scanning root is the current folder, `.`.
+Scanning roots are folders, individual files, or named pipes that should be
+expanded into a list of regular files to be analyzed. Symbolic links are
+allowed as scanning roots.
+
+Expanding a folder consists of listing its contents recursively with
+the following exceptions:
+
+* Symbolic links other than the original scanning roots are ignored.
+* In Git projects, Git submodules are ignored.
+* Paths excluded via Semgrepignore patterns are ignored. Semgrepignore
+  patterns can be of different sources which are detailed in the
+  upcoming section.
+
+The list of files obtained by expanding the scanning roots are called
+**target files**. To obtain target files, Semgrep follows a
+number of fixed rules and some configurable filters.
+
+For each scanning root, Semgrep infers a **project root** (v2 only). The
+project root determines the location of applicable `.semgrepignore`
+files as well as `.gitignore` files in Git projects. In v1 where is no
+notion of a project root, the `.semgrepignore` file is unique and
+looked up in the current folder.
+
+Semgrep determines the project root for each scanning root by first
+obtaining the real path (physical path) to the scanning root. Then,
+Semgrep searches up the file hierarchy for a `.git` folder or
+similar used by one of the popular file version control systems
+(Git, Mercurial, etc.) indicating a project root.
+If no project root is found this way, it
+defaults to the scanning root itself if it is a folder or to its containing
+folder if it is a regular file.
+
+<!-- TODO: explain project detection.
+     Go over options to disable listing files using `git ls-files`
+     while possibly still consulting the `.gitignore` files -- when we
+     have an option for it. Right now we have only `--no-git-ignore`
+     which is confusing and too coarse. I'd like to deprecate it as
+     soon as we have finer-grained replacements.
+-->
+
+:::caution
+As an experimental debugging aid, Semgrep provides the `--x-ls` option
+to list the target files. `--x-ls-long` additionally prints excluded
+files and a brief justification. Beware that these two options are
+likely to be renamed or change their behavior in the
+future. Meanwhile, its typical usage is:
+```
+semgrep --x-ls
+```
+or
+```
+semgrep --x-ls --experimental
+```
+:::
+
+## Sources of Semgrepignore patterns
+
+A Semgrepignore pattern is a glob pattern that is matched by Semgrep
+against file paths to determine whether these paths should be allowed or
+disallowed as target files.
+
+Supported sources of Semgrepignore patterns are:
```

Comment:
From the user perspective, I find "sources" to be an odd word here. They feel like uses to me---places that accept Semgrepignore patterns

Reply:
I rephrased this particular sentence. The term "sources" still occurs before and after this.